vars/cloudFoundryDeploy.groovy

```diff
@@ -0,0 +1,161 @@
+import com.sap.piper.Utils
+import com.sap.piper.ConfigurationHelper
+
+import groovy.transform.Field
+
+@Field String STEP_NAME = 'cloudFoundryDeploy'
+@Field Set STEP_CONFIG_KEYS = [
+    'cloudFoundry',
+    'deployUser',
+    'deployTool',
+    'deployType',
+    'dockerImage',
+    'dockerWorkspace',
+    'mtaDeployParameters',
+    'mtaExtensionDescriptor',
+    'mtaPath',
+    'smokeTestScript',
+    'smokeTestStatusCode',
+    'stashContent']
+@Field Map CONFIG_KEY_COMPATIBILITY = [cloudFoundry: [apiEndpoint: 'cfApiEndpoint', appName:'cfAppName', credentialsId: 'cfCredentialsId', manifest: 'cfManifest', org: 'cfOrg', space: 'cfSpace']]
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+def call(Map parameters = [:]) {
+
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: parameters) {
+
+        def utils = parameters.juStabUtils
+        if (utils == null) {
+            utils = new Utils()
+        }
+
+        def script = parameters.script
+        if (script == null)
+            script = [commonPipelineEnvironment: commonPipelineEnvironment]
+
+        Map config = ConfigurationHelper
+            .loadStepDefaults(this)
+            .mixinGeneralConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS, this, CONFIG_KEY_COMPATIBILITY)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS, this, CONFIG_KEY_COMPATIBILITY)
+            .mixinStageConfig(script.commonPipelineEnvironment, parameters.stageName?:env.STAGE_NAME, STEP_CONFIG_KEYS, this, CONFIG_KEY_COMPATIBILITY)
+            .mixin(parameters, PARAMETER_KEYS, this, CONFIG_KEY_COMPATIBILITY)
+            .dependingOn('deployTool').mixin('dockerImage')
+            .dependingOn('deployTool').mixin('dockerWorkspace')
+            .withMandatoryProperty('cloudFoundry/org')
+            .withMandatoryProperty('cloudFoundry/space')
+            .use()
+
+        echo "[${STEP_NAME}] General parameters: deployTool=${config.deployTool}, deployType=${config.deployType}, cfApiEndpoint=${config.cloudFoundry.apiEndpoint}, cfOrg=${config.cloudFoundry.org}, cfSpace=${config.cloudFoundry.space}, cfCredentialsId=${config.cloudFoundry.credentialsId}, deployUser=${config.deployUser}"
+
+        utils.unstash 'deployDescriptor'
```

Comment:
Maybe also a more general topic. Do we want to stash and unstash in the steps?
Until now we stash and unstash per stage. Thus stashing within a step is not needed. 
Furthermore, if we just consume this step, there is nothing stashed with the name deployDescriptor.

Reply:
Which leads to a red stripe in our pipeline when using this step, without anything wrong from our side. This is not a good user experience, as the users will wonder what went wrong here.

![image](https://user-images.githubusercontent.com/2292245/43253123-7d9003e6-90c4-11e8-931d-69fd6524fdf2.png)

Reply:
We also use stashing massively. However, in for approach it is more a concept of the pipeline. The stages are responsible to load (via unstashing) what is needed in the stage. 
@marcusholl I think we all have to align here, whether stashing should be the responsibility of the steps. I could also image that I want to use the library without stashing or having a different stashing approach. 
It is really bad that the unstashing is marked as red in the pipeline, although everything is fine (as shown by @fwilhe).

Reply:
I agree, it is not nice but I consider this rather a bug in the Blue Ocean UI since we execute the unstash and catch the corresponding exception, Thus it is not an error and should not be displayed like that.
The overall stage is not impacted and most users won't even realize - at least that is the impression I have from our internal usage.

The stashing implemented here which is also the same as used in our internal pipelines is basically agnostic:

* if the stash is there it is used
* if it is not there the existing workspace is used

Thus, to me also having a stashing concept in stages or in the Pipeline is no contradiction, e.g. like in the S/4 SDK pipeline stages.

Since we have this behavior in other steps already, I suggest to leave this discussion out of this PR.